Make storage access throw for meta tensors

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53747 Add 'noarch' tests which only run in one CI config
* **#53681 Make storage access throw for meta tensors**

Without throwing, we can easily segfault trying to access nullptr
storage.

To do this I made set_storage_access_should_throw public so that you
don't have to subclass TensorImpl to do it.  An alternative is
to just bite the bullet and add a MetaTensorImpl subclass.  Let
me know what is preferred.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955540](https://our.internmc.facebook.com/intern/diff/D26955540)

ezyang committed:
Make storage access throw for meta tensors

Without throwing, we can easily segfault trying to access nullptr
storage.

To do this I made set_storage_access_should_throw public so that you
don't have to subclass TensorImpl to do it.  An alternative is
to just bite the bullet and add a MetaTensorImpl subclass.  Let
me know what is preferred.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make storage access throw for meta tensors"


Without throwing, we can easily segfault trying to access nullptr
storage.

To do this I made set_storage_access_should_throw public so that you
don't have to subclass TensorImpl to do it.  An alternative is
to just bite the bullet and add a MetaTensorImpl subclass.  Let
me know what is preferred.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bhosmer reviewed (approved):
This seems fine as a lightweight addition to the public API, since it can only go in the unsafe->safe direction. 

Is a `MetaTensorImpl` inevitable and you just want to add safety in the meantime? Or are there reasons to avoid the subclass permanently?

ezyang commented:
> Is a MetaTensorImpl inevitable and you just want to add safety in the meantime? Or are there reasons to avoid the subclass permanently?

It's not, in fact, I cannot think of any reason why I need to actually make a subclass and would prefer not to.

ezyang committed:
Update on "Make storage access throw for meta tensors"


Without throwing, we can easily segfault trying to access nullptr
storage.

To do this I made set_storage_access_should_throw public so that you
don't have to subclass TensorImpl to do it.  An alternative is
to just bite the bullet and add a MetaTensorImpl subclass.  Let
me know what is preferred.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955540](https://our.internmc.facebook.com/intern/diff/D26955540)

[ghstack-poisoned]

ezyang committed:
Update on "Make storage access throw for meta tensors"


Without throwing, we can easily segfault trying to access nullptr
storage.

To do this I made set_storage_access_should_throw public so that you
don't have to subclass TensorImpl to do it.  An alternative is
to just bite the bullet and add a MetaTensorImpl subclass.  Let
me know what is preferred.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955540](https://our.internmc.facebook.com/intern/diff/D26955540)

[ghstack-poisoned]

ezyang committed:
Update on "Make storage access throw for meta tensors"


Without throwing, we can easily segfault trying to access nullptr
storage.

To do this I made set_storage_access_should_throw public so that you
don't have to subclass TensorImpl to do it.  An alternative is
to just bite the bullet and add a MetaTensorImpl subclass.  Let
me know what is preferred.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26955540](https://our.internmc.facebook.com/intern/diff/D26955540)

[ghstack-poisoned]
